cpp/ql/src/semmle/code/cpp/rangeanalysis/SignAnalysis.qll

```diff
@@ -203,6 +210,8 @@ private predicate unknownSign(Instruction i) {
     i instanceof BuiltInInstruction
     or
     i instanceof CallInstruction
+    or
+    i instanceof ChiInstruction
```

Comment:
What's the correct way to handle `ChiInstruction` in this analysis? I'd expect that chi and pi instructions should be treated similarly.

Reply:
To be sound, you can't make any assumptions about the sign of the result of a `Chi` instruction. First, the result of the `Chi` instruction is a definition of the entire virtual variable, which may include multiple disjoint memory locations as well as multiple fields within the same memory location. Second, the result of a `Chi` instruction is some combination of its two inputs, potentially involving a partial overlap. This is in contrast to a `Phi` instruction, where the result is exactly one of the inputs.

The insertion of `Chi` instructions in this PR is very conservative. It inserts a `Chi` instruction after any partial write to a virtual variable, and hooks up any subsequent use of that virtual variable to the result of the `Chi` instruction, regardless of whether the use overlaps the original definition. A near-future improvement to this will be to pay more attention to the overlap relationship between the original definition and the use. If the definition exactly overlaps that particular use, the use can use the original definition directly instead of using the `Chi` result. If there is no overlap between the original definition and the use, the use can be connected to the nearest earlier definition of that virtual variable that actually does overlap. For example:

```
int x = 2;
int y = 4;
int* p = b ? &x : &y;  // Both x and y are escaped, and so get assigned to the "aliased memory" virtual variable (call it "am")
x = 6;
y = 8;
z = x;
return *p;
```
Building aliased SSA, this becomes:
```
am@1 = AliasedDefinition();
int x@2 = 2;
am@3 = Chi(am@1, x@2);
int y@4 = 4;
am@5 = Chi(am@3, y@4);
int* p = b ? &x : &y;  // Both x and y are escaped, and so get assigned to the "aliased memory" virtual variable (call it "am")
x@6 = 6;
am@7 = Chi(am@5, x@6);
y@8 = 8;
am@9 = Chi(am@7, y@8);
z = x@6;  // Skips am@9 and y@8 because x does not overlap y. Uses x@6 directly.
return *p@9;  // Uses am@9 because *p may or may not overlap y.
```